pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -33,26 +34,71 @@ mod benchmarks;
 pub mod weights;
 pub use weights::WeightInfo;
 
+use tp_traits::{AuthorNotingHook, BlockNumber};
 use {
     dp_core::ParaId,
     frame_support::{
+        dispatch::GetDispatchInfo,
         pallet_prelude::*,
         traits::fungible::{Balanced, Inspect},
     },
     frame_system::pallet_prelude::*,
     parity_scale_codec::EncodeLike,
     sp_runtime::traits::{AccountIdConversion, Convert, Get},
     sp_std::{vec, vec::Vec},
+    staging_xcm::v3::Response,
     staging_xcm::{
         prelude::*,
         v3::{InteriorMultiLocation, MultiAsset, MultiAssets, Xcm},
     },
     tp_traits::ParathreadParams,
 };
 
+pub trait XCMNotifier<T: Config> {
+    fn new_notify_query(
+        responder: impl Into<MultiLocation>,
+        notify: impl Into<<T as Config>::RuntimeCall>,
+        timeout: BlockNumberFor<T>,
+        match_querier: impl Into<MultiLocation>,
+    ) -> u64;
+}
+
+/// Dummy implementation. Should only be used for testing.
+impl<T: Config> XCMNotifier<T> for () {
+    fn new_notify_query(
+        _responder: impl Into<MultiLocation>,
+        _notify: impl Into<<T as Config>::RuntimeCall>,
+        _timeout: BlockNumberFor<T>,
+        _match_querier: impl Into<MultiLocation>,
+    ) -> u64 {
+        0
+    }
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(RuntimeDebug, PartialEq, Eq, Encode, Decode, Clone, TypeInfo)]
+pub struct InFlightCoreBuyingOrder<BN> {
+    para_id: ParaId,
+    query_id: QueryId,
+    ttl: BN,
+}
+
+impl<T: Config> AuthorNotingHook<T::AccountId> for Pallet<T> {
+    fn on_container_author_noted(
+        _author: &T::AccountId,
+        _block_number: BlockNumber,
+        para_id: ParaId,
+    ) -> Weight {
+        PendingBlocks::<T>::remove(para_id);
```

Comment:
Does this imply a write even if the pendingBlocks does not exist? I am saying this because it is likely that para_ids for **parachains** and not **parathreads** are not present here. Yet we are adding a write for every single one of them

Reply:
I would expect `remove()` to be more efficient than `if exists() { remove() }` in terms of collator cpu time, but not sure what happens to PoV if you remove a storage item that does not exist.

Reply:
the point is that you are going to call this probably for parachains and parathreads. And yes for parathreads is faster, but not for parachains.

Reply:
PoV should not increase in size. PoV is only consist of Markle trie nodes that are being accessed (read) during the block execution.

Reply:
@girazoki I went through the entire code path. 

1. For directly calling `remove()`  the change is first written to in-memory overlayed changes data structure  of the runtime api instance from which at the time of block building we extract storage changes. These storage changes  are then applied atomically to the current state of the chain to produce block. Here, advantage of direct removal is that PoV does not increase but we do have additional entry in the overlayed changes which is internal struct and only exists in memory between the time extrinsics are executed to the state is modified (i.e block is imported.).

2. Regarding checking if the key exists or not. This becomes part of the runtime api's `StorageProof` data structure which is used in building PoV.  I believe CPU/memory overhead of local node will be same in both cases but checking existence mean that we are reading something from the storage so it will be recorded by proof recorder and will become part of the storage proof and consequently of the `PoV`. Which I believe is much bigger concern than extra entry in overlaid changes.

So, my opinion would be to directly call `remove`

Reply:
I guess in the future we should add a `is_parathread: bool` parameter to `on_container_author_noted`, because it is used in pallets inflation-rewards and service-payment, and those will probably be different for parachains vs parathreads.